src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,376 @@
+            i += 1
+
+        if not strategies:
+            return hs.just(EMPTY)
+        if len(strategies) == 1:
+            return strategies[0]
+        return hs.tuples(*strategies).map(EMPTY.join)
+    else:
+        # Single code
+        code, value = codes
+        if code == sre.LITERAL:
+            # Regex 'a' (single char)
+            c = to_char(value)
+            if context.flags & re.IGNORECASE:
+                return hs.sampled_from([c.lower(), c.upper()])
+            return hs.just(c)
+
+        elif code == sre.NOT_LITERAL:
+            # Regex '[^a]' (negation of a single char)
+            c = to_char(value)
+            blacklist = set([c.lower(), c.upper()]) \
+                if context.flags & re.IGNORECASE else [c]
+            if is_text:
+                return hs.characters(blacklist_characters=blacklist)
+            else:
+                return hs.binary().filter(lambda c: c not in blacklist)
+
+        elif code == sre.IN:
+            # Regex '[abc0-9]' (set of characters)
+            charsets = value
+
+            builder = CharactersBuilder(negate=charsets[0][0] == sre.NEGATE,
+                                        flags=context.flags)
+
+            for charset_code, charset_value in charsets:
+                if charset_code == sre.NEGATE:
+                    # Regex '[^...]' (negation)
+                    pass
+                elif charset_code == sre.LITERAL:
+                    # Regex '[a]' (single char)
+                    builder.add_chars(hunichr(charset_value))
+                elif charset_code == sre.RANGE:
+                    # Regex '[a-z]' (char range)
+                    low, high = charset_value
+                    for char_code in hrange(low, high + 1):
+                        builder.add_chars(hunichr(char_code))
+                elif charset_code == sre.CATEGORY:
+                    # Regex '[\w]' (char category)
+                    builder.add_category(charset_value)
+                else:  # pragma: no cover
+                    # Currently there are no known code points other than
+                    # handled here. This code is just future proofing
+                    raise InvalidArgument(
+                        'Unknown charset code: %s' % charset_code
+                    )
+
+            if not is_text:
+                return builder.strategy.map(lambda u: u.encode('utf-8'))
+            return builder.strategy
+
+        elif code == sre.ANY:
+            # Regex '.' (any char)
+            if is_text:
+                if context.flags & re.DOTALL:
+                    return hs.characters()
+                return hs.characters(blacklist_characters=u'\n')
+            else:
+                if context.flags & re.DOTALL:
+                    return hs.binary()
+                return hs.binary().filter(lambda c: c != b'\n')
+
+        elif code == sre.AT:
+            # Regexes like '^...', '...$', '\bfoo', '\Bfoo'
+            if value == sre.AT_END:
```

Comment:
I'll correct the docs - `^a$` may match `a` or `a\n`.  The important part is that we don't notice `a$b` until filtering, but that's an implementation issue.